rust-bindings/src/lib.rs

```diff
@@ -0,0 +1,41 @@
+#![allow(non_camel_case_types)]
+#![allow(non_snake_case)]
+
+mod bindings {
+    include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
+}
+
+/// Encode a BIP158 filter from a list of slices.
+pub fn encode_filter(slices: &[&[u8]]) -> Box<[u8]> {
+    // Convert the slices to a type that C/C++ can understand.
+    let slices: Vec<bindings::Slice> = slices
+        .iter()
+        .map(|slice| bindings::Slice {
+            bytes: slice.as_ptr(),
+            length: slice.len(),
+        })
+        .collect();
+
+    // SAFETY: The length provided matches the length of the slice, so this should be safe.
+    unsafe {
+        // Call the binding to generate the BIP158 filter.
+        let filter = bindings::encode_filter(slices.as_ptr(), slices.len());
+
+        // Convert the C `Slice` struct to a standard Rust slice.
+        let slice = std::slice::from_raw_parts(filter.bytes, filter.length);
+
+        // We cast the slice to a `Box<[u8]>` to ensure that it is properly deallocated.
+        // The slice is declared as `const` in the C struct, but nothing else uses it so it should be safe to make mutable.
+        // And then take ownership with `Box` since it's on the heap and must be freed later.
+        Box::from_raw(slice as *const [u8] as *mut [u8])
```

Comment:
is `Box` really guaranteed to use a compatible way of freeing memory?
I would not think so. From C++'s point of view, the memory is allocated with `new[]` and must be freed by `delete[]`. I bet rust frees it with `free()`, which may work on some or most platforms, but is breaking the rules and may fail on some other platforms. Windows may use different heaps for instance.

This is from the documentation of `from_raw()`:

> For this to be safe, the memory must have been allocated in accordance with the [memory layout](https://doc.rust-lang.org/std/boxed/index.html#memory-layout) used by Box .